chore: Add pytest and selenium test framework to the project

m1yag1 opened the pull request:
Added pytest and selenium test framework to the project

m1yag1 committed:
Added Python and pytest specific items to the .gitignore

m1yag1 committed:
Added black code formatter and pre-commit-configuration to the project

m1yag1 committed:
added a .env.example file to the project

m1yag1 committed:
Added a Makefile-py for Python related functions

* The Makefile can create a python virtualenv in the local directory and
install dependencies.
* The Makefile can clean any build aritifacts or test artifacts related
to Python.
* The Makefile can clean any bytecode files created by Python.

m1yag1 committed:
added Python requirements-dev.txt file to the project

m1yag1 committed:
added basic python directory structure and prelim files

m1yag1 committed:
added testrail.example.cfg

m1yag1 committed:
updated the README to oint to ui tests documentation

m1yag1 committed:
fixed link to pytest tests/ui/README.md in root README.md file

m1yag1 committed:
Updated the tests/ui/README.md to be more rex-web specific

TomWoodward commented on .gitignore:
```diff
@@ -27,3 +27,100 @@ yarn-error.log*
 /localhost_*.report.html
+
+# Jetbrains files
+.idea/
```
i'm not morally opposed to ignoring editor files in the project gitignore, but since your editor is a personal decision, you could add them to your [global gitignore](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore)

m1yag1 replied:
I can put it in my global gitignore. no biggie.

m1yag1 replied:
Also, .DS_Store is in there.. i had added it again on accident, but that would also apply to your comment too. That's a OSX specific thing and operating systems to develop are a personal choice too ;).

TomWoodward commented on .gitignore:
```diff
@@ -27,3 +27,100 @@ yarn-error.log*
+*.sage.py
+
+# Environments
+.env
```
`.env` is intentionally allowed, use `.env.local` for stuff outside version control. that said we don't use the environment files for the JS on this project anymore so if you want to leave this you can. if you do leave it maybe combine it with all the env related stuff above

m1yag1 replied:
Ah, that's a good practice. I'll have to start doing that with the other projects we have. 👍

TomWoodward commented on .gitignore:
```diff
@@ -27,3 +27,100 @@ yarn-error.log*
+nosetests.xml
+coverage.xml
+*.cover
+.hypothesis/
```
i feel as if some of this stuff isn't necessary

m1yag1 replied:
Yea, again, it's the cause of some copy pasta.. I'll remove it.

TomWoodward commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,12 @@
+# See https://pre-commit.com for more information
+# See https://pre-commit.com/hooks.html for more hooks
+repos:
```
@philschatz maybe we should add the faster checks like `tsc` and `tslint` as commit hooks?

m1yag1 replied:
I really like the use of pre-commit hooks. I think they are a great way to run the checks as you commit instead of waiting for some other time. Even though I've had linters run based on the test command I always seem to break CI for some dumb linting error. Now that I use pre-commit hooks it doesn't happen anymore.

TomWoodward replied:
i would definitely not want a pre-commit hook to run tests because i don't want to wait for them every time i commit. the linter is pretty fast though. i like to have control over when my code gets re-formatted and would not love to have it happen automatically on commit

m1yag1 replied:
yea, i don't run tests.. i don't like that. only linting checks on the code i'm committing. We can just use this for the python stuff. I can move it to the python Makefile.

TomWoodward commented on Makefile-py:
```diff
@@ -0,0 +1,37 @@
+
```
can this be in the folder with all the python stuff? or rename it to just `Makefile` and i'll start adding rando scripts here too

m1yag1 replied:
My thought was that if other Make commands were introduced you could create one Make file that combines them all. It can be useful to separate out the docker specific Make stuff using something like `make -f Makefile-docker` but also combine them with a core Makefile and use `include Makefile-docker`. I thought the separation might help but maybe not. 🤷‍♂️

TomWoodward replied:
having sub-makefiles and composing them sounds like a cool pattern, in that case could we move the Makefile-py into a python subdirectory? or would that mess with the cwd of the scripts and be confusing? i'm not _opposed_ to having stuff in the top level, i'd just like to put things in subdirectories that can be.

m1yag1 replied:
yes we could move the py Makefile to a subdirectory.

m1yag1 replied:
I moved it. This is resolved.

TomWoodward commented on pyproject.toml:
```diff
@@ -0,0 +1,21 @@
+[tool.black]
```
if we could move as many of the python project files into a subdirectory as possible that would make me happy

m1yag1 replied:
Sure, I explored this at first but there were a few things that looked to the root directory for configs (the big one is black which is the code formatter for python) and the work arounds were like, "create a bash script that automatically adds the --config-file /some/path/to/config". My main idea was to go for easy to run w/o having someone remember a bunch of command line flags for config files.

Additionally, I've worked in a number of projects that use python and js. I don't think I've ever been like, "let me shove the node config stuff into another folder b/c it's not python." The root allows one to see the files necessary to run the project. However, I do understand the need to keep the root clean and only have files necessary for the project's runtime. An example of python and js living happily together is here: https://github.com/openstax/research-eGrader

I'll try to get things working where python is hidden. That may require more time investment than I have this week so I may not finish this soon as next week I have other things I need to work on.

m1yag1 replied:
oh hypothesis is another example that has js stuff in the root along with the python: https://github.com/hypothesis/h

TomWoodward replied:
yea, again, not opposed. the idea is that we would be working in the py tests regularly (for almost every feature developed probably) and i'm not trying to make the py code second class, even for the ts files i push them down the tree as much as i can

m1yag1 replied:
I was able to move most of the files into the tests directory. The only one I wasn't able to is the config for pre-commit.

TomWoodward commented on testrail.example.cfg:
```diff
@@ -0,0 +1,10 @@
+[API]
+url = https://instance.testrail.net/
```
i know this is just an example, but in reality we're going to have to eventually pull all these configs from the environment, is that possible?

m1yag1 replied:
Hmm.. I don't think the pytest-testrail plugin uses env vars. I'll have to double check.

m1yag1 replied:
Yea, pytest-rail out of the box doesn't support environment variables. I'll need to add those.

TomWoodward replied:
ok, in CI we can specify them via command line, that'll be easier

m1yag1 replied:
👍

m1yag1 replied:
I removed the file and added instructions in the readme to create a file if someone wanted to run it locally and create their own run.

TomWoodward commented on tests/ui/README.md:
```diff
@@ -0,0 +1,153 @@
+
+## Uploading results to TestRail
+
+The TestRail integration is currently intended to be used during a local test run of the cnx-automation suite when the uploading of results to TestRail is desired.
```
rex-web

m1yag1 replied:
Fixed!

TomWoodward reviewed (commented):
what is there so far of the page structure and framework looks really good. i think all my concerns were about merging the py and ts project configs at the top level of the repo. it would make me feel better to move the python specific configs into a subdirectory

m1yag1 committed:
remove unneeded items from .gitignore and remove personal choices.

m1yag1 committed:
Added content.py page with toc region along with basic toc tests

* Added a Base page (`pages/base.py`) to be used for other pages.
* Added a Base region (`regions/base.py`) to be used for regions.
* Added the `content.py` page object with a TableOfContents Region.
* Added a chapters property to the TableOfContents Region.
* Added locators and properties in the Content page for various things needed for the
finding the toc toggle button and clicking on it.
* Added a test file for table of contents called `test_toc`
* Added a basic test for the toc button toggle.
* Added a basic to check for chapters.
* Moved the flake8 config file to the appropriate place ;).
* Added a markers file for more simplicity and standardization.
`@markers.parametrize()` can be used instead of
`pytest.mark.paramentrize` and then other variations.

m1yag1 committed:
update THEN for test_toc_has_chapters test

m1yag1 commented on tests/ui/test_toc.py:
```diff
@@ -0,0 +1,35 @@
+
+@markers.parametrize("book_slug,page_slug", [("college-physics", "preface")])
+@markers.nondestructive
+def test_toc_toggle_button_opens_and_closes(selenium, base_url, book_slug, page_slug):
```
@TomWoodward Here I added a basic test of clicking the TOC toggle button. This test currently passes for both desktop and mobile resolutions. However, in the mobile view it really covers up the screen so I would suggest that the mobile table of contents be hidden by default.

TomWoodward replied:
i believe this is the plan you can see the upcoming new designs here openstax/unified-issues#104

m1yag1 committed:
Create main Makefile and move Makefile-py into the  directory

m1yag1 committed:
updated venv requirements path in  and moved files

m1yag1 committed:
deleted testrail.example.cfg

m1yag1 committed:
remove testrail.example file and add instructions if needed in README.md

m1yag1 committed:
Merge branch 'master' of github.com:openstax/rex-web into added-pytest-selenium

TomWoodward commented on tests/ui/pages/base.py:
```diff
@@ -0,0 +1,84 @@
+            raise IndexError("Unknown dimension: {}".format(dimension))
+        return get_size[dimension]
+
+    def set_window_size(self, width=0, height=0):
```
in javascript i would create a util decorator function that did the size guard and sleep that you just pass a function reference for `set_window_size` or `set_window_position` into.

i'm not sure if there is a comparably easy thing to do in python, but these two blocks of code are nearly identical.

m1yag1 replied:
I removed this as it's probably better to implement when we need it.

TomWoodward commented on tests/ui/pages/base.py:
```diff
@@ -0,0 +1,84 @@
+        """Attempt to change the browser window size."""
+        if width >= 1 and height >= 1:
+            self.driver.set_window_size(width, height)
+            sleep(1.0)
```
is there no way to check the window size to verify this worked instead of arbitrarily sleeping?

m1yag1 replied:
I removed this code like the previous code. This was done for some reason like it b/c of something with cnx-automation. we probably don't need to bring that code over unless the specifics of the test are similar.

TomWoodward commented on tests/ui/pages/content.py:
```diff
@@ -0,0 +1,50 @@
+        '#main-content > div > div > div h1[data-type="document-title"]',
+    )
+    _table_of_contents_button_locator = (By.CSS_SELECTOR, "#root > div > div > button")
+    _table_of_contents_nav_locator = (By.CSS_SELECTOR, "#root > div > div > div > div > nav")
```
@philschatz should weigh in but i think we should just sprinkle some `aria-label`s on these to select them by

m1yag1 replied:
I agree. They didn't exist so this is what i did to get the tests to identify the right elements.

TomWoodward commented on tests/ui/pages/content.py:
```diff
@@ -0,0 +1,50 @@
+
+    @property
+    def loaded(self):
+        return self.find_element(*self._main_content_locator)
```
we should really use the js hook that the app provides, that is the canonical way that we're identifying when the page is done doing _stuff_

TomWoodward replied:
when we run these tests against a built environment the page will be complete before the js is even loaded, so then this code would thought the page was done loading but all interactions would fail

m1yag1 replied:
Could you provide me with some js code that would allow me to resolve the js hook. It would look something like this.

https://github.com/openstax/cnx-automation/blob/try-applitools/tests/webview/ui/test_applitools.py#L25-L27

I have no idea what the return on something like that is.. i would need to experiment.

TomWoodward commented on tests/ui/regions/base.py:
```diff
@@ -0,0 +1,59 @@
+        """
+        return self.page.scroll_up()
+
+    def scroll_to(self, element=None):
```
similar to the other situation, i'd consider changing this to `return self.page.scroll_to(element or self.root)` or `return self.page.scroll_to(defaultElement(element))`

m1yag1 committed:
removed all unused code not relevant to what is being tested.

m1yag1 committed:
replace webview with rex-web

TomWoodward commented:
@m1yag1 when i try to run `pytest --base-url http://localhost:3000 tests/ui` i get 

```bash
============================= test session starts ==============================
platform darwin -- Python 3.7.0, pytest-4.1.1, py-1.7.0, pluggy-0.8.1 -- /Users/tomwoodward/code/openstax/books-web/.venv/bin/python3
cachedir: .pytest_cache
sensitiveurl: ^(?:https?\:\/\/)?openstax\.org
metadata: {'Python': '3.7.0', 'Platform': 'Darwin-17.7.0-x86_64-i386-64bit', 'Packages': {'pytest': '4.1.1', 'py': '1.7.0', 'pluggy': '0.8.1'}, 'Plugins': {'variables': '1.7.1', 'selenium': '1.15.1', 'metadata': '1.8.0', 'html': '1.20.0', 'base-url': '1.4.1'}, 'Base URL': '', 'Driver': None, 'Capabilities': {}}
rootdir: /Users/tomwoodward/code/openstax/books-web/tests/ui, inifile: setup.cfg
plugins: variables-1.7.1, selenium-1.15.1, metadata-1.8.0, html-1.20.0, base-url-1.4.1
collecting ... collected 0 items / 1 errors

==================================== ERRORS ====================================
_________________________ ERROR collecting test_toc.py _________________________
ImportError while importing test module '/Users/tomwoodward/code/openstax/books-web/tests/ui/test_toc.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/ui/test_toc.py:1: in <module>
    from tests.ui.pages.content import Content
E   ModuleNotFoundError: No module named 'tests'
- generated html file: /Users/tomwoodward/code/openstax/books-web/ui-test.html -
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.17 seconds ============================
```

i haven't installed the webdriver stuff yet, but this doesn't look related i don't think...

TomWoodward committed:
fix typo in readme

TomWoodward commented on README.md:
```diff
@@ -97,14 +97,13 @@ this suite:
 ### Selenium Tests
 
-this suite has not been implemented yet.
+run these with `pytest --base-url http://localhost:3000 tests/ui`
```
all of our other tests are distributed in the js code next to the stuff that they're testing. we have a /src/test place for making testing utilities, mocks, etc.

because of this, i don't think we're ever going to want to put anything else in the top level /tests folder, and having it wrap the single ui folder seems a little weird. can we rename the whole thing to /selenium or /ui-tests or something like that?

m1yag1 replied:
I figured this would come up and just picked tests as that was standard for python. I also wasn't sure if we would employ python to do some other kind of testing like integration tests... so i split it up like that... thinking it may be `tests(or something else)/ui`,`tests/integration`, `tests/visual`

maybe just name the root folder to `pytest`?

m1yag1 replied:
I based it off what mozilla had. https://github.com/mozilla/addons-server/tree/master/tests

TomWoodward replied:
👍 for root folder `pytest`, if you think its prudent to make a subfolder below that go for it

m1yag1 committed:
Changed the pytest folder structure

* Removed `ui` folder and renamed `tests` folder to `pytest-selenium`
* Moved all files from `ui` folder into the root of `pytest-selenium`
* Updated Makefile-py and main Makefile to point to the right submake
* Created a tests folder inside pytest-selenium and moved conftest.py and the `tests_toc.py`
file into that folder.`
* Made pages and regions folders into py packages by adding
`__init__.py` to each folder.
* Updated imports to get tests running with the new structure.

m1yag1 committed:
Merge branch 'added-pytest-selenium' of github.com:openstax/rex-web into added-pytest-selenium

m1yag1 committed:
update main README.md with pytest command

TomWoodward commented on Makefile:
```diff
@@ -0,0 +1,14 @@
+.PHONY: help pre-commit clean clean-build clean-pyc clean-test venv
```
this line is also in the py makefile, does it need to have the same commands in both places?

m1yag1 replied:
i believe so. if someone were to do `make -f pytest-selenium/Makefile-py`

m1yag1 replied:
However, that is not likely so i'll remove it ;)

TomWoodward commented on pytest-selenium/README.md:
```diff
@@ -0,0 +1,151 @@
+
+The TestRail integration is currently intended to be used during a local test run of the rex-web pytest suite when the uploading of results to TestRail is desired.
+
+Make a copy of of the testrail.example.cfg:
```
this doesn't exist

TomWoodward commented on pytest-selenium/tests/conftest.py:
```diff
@@ -0,0 +1,77 @@
+
+
+@pytest.fixture(
+    scope="function", params=[DESKTOP, MOBILE], ids=["Resolution: 1080x1920", "Resolution: 414x738"]
```
resolutions duplicated here and above?

m1yag1 replied:
The ids are mainly for display when the tests are running. It's not actually duplicating them but rather adding a label.

m1yag1 replied:
Example:
![image](https://user-images.githubusercontent.com/8730430/51543662-b1f7b380-1e23-11e9-8de3-4b16717f4015.png)

TomWoodward replied:
but why not use the constants in these strings?

m1yag1 replied:
I did

m1yag1 committed:
remove PHONY line in pytest-selenium/Makefile-py

m1yag1 committed:
Fix instruction to create testrail.cfg file in pytest-selenium/README.md

m1yag1 commented on README.md:
```diff
@@ -97,14 +97,13 @@ this suite:
 ### Selenium Tests
 
-this suite has not been implemented yet.
+run these with `pytest --base-url http://localhost:3000 --driver pytest-selenium/tests`
```
Haha, yea you're right

m1yag1 replied:
yea sorry.. I typically pass in Chrome as a default. Since we have a focus on cross browser it's something more explicit. I'm not used to it having to be passed in manually.

m1yag1 commented on pytest-selenium/README.md:
````diff
@@ -0,0 +1,151 @@
+To run a specific test, pass in a value for `-k`:
+
+```bash
+$ pytest -k test_my_feature tests/ui
````
Yes, a lot of these have changed with the new changes. I'll correct them.

m1yag1 committed:
Make changes to pytest run commands using new folder structure and including --driver

m1yag1 committed:
deleted .env.example as it's not currently being used

m1yag1 committed:
added pytest-testrail plugin and marker in markers.py

TomWoodward committed:
Add testrail marker for toc test case

TomWoodward committed:
Merge branch 'added-pytest-selenium' of github.com:openstax/books-web into added-pytest-selenium

TomWoodward committed:
start work on test command

TomWoodward committed:
update test command

TomWoodward committed:
use a real rex url

TomWoodward committed:
Merge branch 'master' into added-pytest-selenium

TomWoodward reviewed (dismissed):
this is working both locally and through saucelabs and can post the results back to testrail. i'll make some additional changes when i try to automate things more (the `make test` command is mostly just note taking for me) but i think we should get this merged down! :shipit:

philschatz reviewed (changes requested):
This is just a review of the precommit hooks, `Makefile`, and `README.md` (awesome to read through!) but will submit another one for the actual code

philschatz commented on pytest-selenium/pages/content.py:
```diff
@@ -0,0 +1,45 @@
+        By.CSS_SELECTOR,
+        '#main-content > div > div > div h1[data-type="document-title"]',
+    )
+    _table_of_contents_button_locator = (By.CSS_SELECTOR, "#root > div > div > button")
```
```suggestion
    _table_of_contents_button_locator = (By.CSS_SELECTOR, "[aria-label='Click to close the Table of Contents']")
```

Relying on the `[aria-label]` rather than a fragile selector helps test accessibility _and_ is less fragile than a `>` selector

philschatz commented on pytest-selenium/pages/content.py:
```diff
@@ -0,0 +1,45 @@
+
+    _main_content_locator = (
+        By.CSS_SELECTOR,
+        '#main-content > div > div > div h1[data-type="document-title"]',
```
What are some reasons for using very-specific selectors in the Selenium tests?

```suggestion
        'h1',
```

If the WCAG guidelines are followed (I think we do so far) there should _only_ be one `h1` in the whole document. And if we don't follow the guidelines then something should blow up anyway 😄

philschatz commented on pytest-selenium/pages/content.py:
```diff
@@ -0,0 +1,45 @@
+
+    @property
+    def loaded(self):
+        return self.find_element(*self._main_content_locator)
```
Does `self.find_element` throw an error when the locator matches `0` or `> 1` elements? If not, would it make sense to add a help that does that?

philschatz replied:
Even when the locator matches `> 1` elements? I'm hoping that selenium can:

- help ensure there is exactly 1 `h1` on the page
- explicitly fail when more than 1 element matches a selector

m1yag1 replied:
We would need to write a test for that. Selenium doesn't enforce any standards like that.

m1yag1 replied:
What could be automated is some sort of HTML validation. I don't know any tools that do that off the top of my head but i'm sure something could be integrating. It would be something very similar to an aXe report. We would have it fail when there were any failures or some threshold.

m1yag1 replied:
Something like this seems like it would do what you want. https://validator.w3.org/docs/api.html

philschatz replied:
(just bike-shedding)

Oh, I guess that I was trying to say that if the test-writer expects the selector to match one element (by using `find_element` rather than `find_elements`) then it seems like that assumption should throw an error if more than one element is matched. Otherwise, `find_element` really means `find_first_element`.

Once I contribute a test that uses `find_element` I might refactor all of the `find_element` references to be something like `util.find_element(self, ...)` or something.

philschatz commented on pytest-selenium/pages/content.py:
```diff
@@ -0,0 +1,45 @@
+        '#main-content > div > div > div h1[data-type="document-title"]',
+    )
+    _table_of_contents_button_locator = (By.CSS_SELECTOR, "#root > div > div > button")
+    _table_of_contents_nav_locator = (By.CSS_SELECTOR, "#root > div > div > div > div > nav")
```
```suggestion
    _table_of_contents_nav_locator = (By.CSS_SELECTOR, "nav")
```

m1yag1 replied:
Did you test this change?

philschatz replied:
eep, there are 36 `<nav>` elements. I think that's a bug... there should only be one.

philschatz replied:
This will probably be fixed to be just one `<nav>` as part of the "Add ToC Toggle" story in "Reading Experience": https://docs.google.com/document/d/1-KX5NMSeJTP5309L7mlXw85YHG_7XRFm-C7ofIdWtCE/edit#heading=h.apfasqkrfa3g

TomWoodward replied:
@philschatz the nav element is used for navs and subnavs, so each chapter has its own nav. i don't know of any reason to limit to just 1 nav per page

philschatz replied:
From the [MDN Description of nav](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/nav) it seems like they are used to denote separate navigations on a page (e.g. ToC, site navigation) and are not nested. In the epub spec the entire ToC is a `<nav>` which contains nested `<ul>`, `<li>`, and `<a href="">` elements.

TomWoodward replied:
ok, we can definitely change it to only be one nav for the ToC, we may end up having another one for site nav though so we should assume there are at least a couple

philschatz replied:
Yeppers, and when a 2nd `<nav>` is introduced they can be differentiated via `aria-label` attributes so the selenium tests _and_ screenreaders can differentiate 😄

philschatz commented on pytest-selenium/pages/content.py:
```diff
@@ -0,0 +1,45 @@
+        return self.TableOfContents(self)
+
+    class TableOfContents(Region):
+        _root_locator = (By.CSS_SELECTOR, "#root > div > div > div > div > nav")
```
Is this different than the `_table_of_contents_nav_locator` ? If not, can they be combined?

philschatz commented on pytest-selenium/setup.cfg:
```diff
@@ -0,0 +1,3 @@
+[tool:pytest]
+addopts=-vs --tb=long --showlocals --html ui-test.html --self-contained-html
+sensitive_url=^(?:https?\:\/\/)?openstax\.org
```
What is this field used for? I don't see it used anywhere directly... maybe it is used by an included python package?

m1yag1 replied:
These are defaults passed into the pytest command. You run pytest --driver Chrome pytest-selenium/tests and these are inserted in there so you don't have to type them out.

philschatz replied:
ah ok. I was just curious about the purpose of `sensitive_url` but thanks for enumerating the others

TomWoodward commented on pytest-selenium/pages/base.py:
```diff
@@ -0,0 +1,24 @@
+from selenium.webdriver.common.keys import Keys
+
+
+class Page(pypom.Page):
```
the reason we're using python in the first place is because mike has already solved this problem using python and doesn't have to re-invent everything to set up rex in the same way. down the line once we get things working i'd consider porting the suite over to typescript. in the meantime i don't think its worth going too far out of our way to modify the base pytest framework (like adding types)

m1yag1 replied:
it will be some time before I can go through and add all the type hints.

philschatz commented on pytest-selenium/tests/conftest.py:
```diff
@@ -0,0 +1,79 @@
+import pytest
+
+# Window resolutions
+DESKTOP = (1080, 1920)
```
Is there a reason these are defined as `height,width` here but they are used in both the code and the docs as `width,height`?

m1yag1 replied:
Selenium uses these backwards from the way we typically see them. I've adjusted them in the documentation so that they use the typical order.

m1yag1 replied:
Actually it might be that pytest injects them in the opposite order.

philschatz commented on pytest-selenium/tests/conftest.py:
```diff
@@ -0,0 +1,79 @@
+        chrome_options.headless = True
+
+    # Required to run in Travis containers
+    if pytestconfig.getoption("--no-sandbox"):
```
Could these be conditional on running in a container (environment `CI=true` or `TRAVIS=true`)? Once we start storing cookies (logged-in, whether popups were dismissed, etc) it might cause tests to fail unexpectedly.

m1yag1 replied:
Sure, it might but I'd rather not make changes on what might happen. This is basically copied from our other automation framework that does run in travis.

philschatz commented on pytest-selenium/tests/test_toc.py:
```diff
@@ -0,0 +1,23 @@
+@markers.test_case('C250849')
+@markers.parametrize("book_slug,page_slug", [("college-physics", "preface")])
+@markers.nondestructive
+def test_toc_toggle_button_opens_and_closes(selenium, base_url, book_slug, page_slug):
```
❤️ The whole test is so sooo concise and readable 😄

philschatz commented on README.md:
```diff
@@ -97,14 +97,13 @@ this suite:
 ### Selenium Tests
 
-this suite has not been implemented yet.
+run these with `pytest --base-url http://localhost:3000 --driver Chrome pytest-selenium/tests`
```
```suggestion
make venv
source ./.venv/bin/activate
run these with `pytest --base-url http://localhost:3000 --driver Chrome pytest-selenium/tests`
```

This _partially_ worked for me but I got the following error (do we need separate instructions for installing chromedriver?):

```
selenium.common.exceptions.WebDriverException: Message: 'chromedriver' executable needs to be in PATH.
```

Maybe it would make sense to add a Makefile target instead? Like `make test-local` or something?

philschatz committed:
add chromedriver to the path

philschatz committed:
adds `make test-local`

to simplify running all the tests against a local instance

philschatz committed:
:shirt: change echo tabs to spaces

(were they intentional?)

philschatz committed:
Update pytest-selenium/README.md

Co-Authored-By: m1yag1 <mike.arbelaez@rice.edu>

m1yag1 commented on pytest-selenium/README.md:
```diff
@@ -0,0 +1,133 @@
+
+Add these configs with the appropriate values:
+
+    [API]
```
Most of the time i believe this will be ran by automation. These instructions are here in the event if someone wants to do that locally which may not be that often.

philschatz committed:
address PR comments

philschatz commented:
@m1yag1 Do you want to make any comments before this gets merged? I pushed a few commits addressing my review comments yesterday but forgot to ask you to review

philschatz committed:
Merge branch 'master' into added-pytest-selenium

m1yag1 committed:
comment on the use of resolution tuples being used by pytest backwards

m1yag1 committed:
Merge branch 'added-pytest-selenium' of github.com:openstax/rex-web into added-pytest-selenium

TomWoodward committed:
Merge branch 'master' into added-pytest-selenium

TomWoodward commented on README.md:
```diff
@@ -97,14 +97,13 @@ this suite:
 ### Selenium Tests
 
-this suite has not been implemented yet.
+run these with `make test-local`
```
this command doesn't seem to exist

m1yag1 commented:
I did a merge of Phil’s code but had a conflict. I thought i put that in there. ugh. i think i messed something up.
On Feb 1, 2019, 9:59 AM -0600, Thomas Woodward <notifications@github.com>, wrote:
> @TomWoodward approved this pull request.
> In README.md:
> > @@ -97,14 +97,13 @@ this suite:
>
> ### Selenium Tests
>
> -this suite has not been implemented yet.
> +run these with `make test-local`
> this command doesn't seem to exist
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

m1yag1 committed:
renamed test to test-local in Makefile and fixed path in README.md

m1yag1 committed:
Merge branch 'added-pytest-selenium' of github.com:openstax/rex-web into added-pytest-selenium

m1yag1 committed:
rename test to test-local in Makefile help_submake command

m1yag1 committed:
Merge branch 'master' into added-pytest-selenium

philschatz committed:
:art: links

philschatz committed:
clarify that make test does not work

philschatz reviewed (dismissed):
I just fixed a link and clarified that `make test` does not work by printing out the help & erroring so people know to run `make test-local` or `make test-rail`

philschatz committed:
Update pytest-selenium/README.md pytest command to have ./ in filepath

Co-Authored-By: m1yag1 <mike.arbelaez@rice.edu>

philschatz commented:
oooh, pyyaml has a security vulnerability & should be upgraded to `pyyaml>=4.2b1`. https://github.com/openstax/rex-web/network/alerts

m1yag1 commented:
Find element returns the first one it finds. If you want more than one you have to use find_elements. I do see where it would be useful.
  

  
>   
> On Feb 1, 2019 at 5:58 PM,  <Philip Schatz (mailto:notifications@github.com)>  wrote:
>   
>   
>   
>
> @philschatz  commented on this pull request.
>
>     
>
> In   pytest-selenium/pages/content.py (https://github.com/openstax/rex-web/pull/68#discussion_r253236423):
>
>   
>   >  +from regions.base import Region + + +class Content(Page): + URL_TEMPLATE = "/books/{book_slug}/pages/{page_slug}" + + _main_content_locator = ( + By.CSS_SELECTOR, + '#main-content  >  div  >  div  >  div h1[data-type="document-title"]', + ) + _table_of_contents_button_locator = (By.CSS_SELECTOR, "#root  >  div  >  div  >  button") + _table_of_contents_nav_locator = (By.CSS_SELECTOR, "#root  >  div  >  div  >  div  >  div  >  nav") + + @property + def loaded(self): + return self.find_element(*self._main_content_locator)  
>   
>
> (just bike-shedding)
>
>   
>
> Oh, I guess that I was trying to say that if the test-writer expects the selector to match one element (by using  find_element  rather than  find_elements) then it seems like that assumption should throw an error if more than one element is matched. Otherwise,  find_element  really means  find_first_element.
>
>   
>
> Once I contribute a test that uses  find_element  I might refactor all of the  find_element  references to be something like  util.find_element(self, ...)  or something.
>
>   
>
> —
>  You are receiving this because you were mentioned.
>  Reply to this email directly,   view it on GitHub (https://github.com/openstax/rex-web/pull/68#discussion_r253236423), or   mute the thread (https://github.com/notifications/unsubscribe-auth/AIU3PuMKAYBNGoiunkROimTrgf8xh4UAks5vJNSdgaJpZM4aEIvL).
>
>     
>
